extensions/barrage/src/main/java/io/deephaven/extensions/barrage/chunk/FixedWidthChunkInputStreamGenerator.java

```diff
@@ -0,0 +1,130 @@
+package io.deephaven.extensions.barrage.chunk;
+
+import gnu.trove.iterator.TLongIterator;
+import io.deephaven.chunk.ObjectChunk;
+import io.deephaven.chunk.WritableLongChunk;
+import io.deephaven.chunk.WritableObjectChunk;
+import io.deephaven.chunk.attributes.Values;
+import io.deephaven.extensions.barrage.util.StreamReaderOptions;
+import io.deephaven.util.datastructures.LongSizedDataStructure;
+
+import java.io.DataInput;
+import java.io.IOException;
+import java.util.Iterator;
+
+import static io.deephaven.util.QueryConstants.NULL_CHAR;
+
+public class FixedWidthChunkInputStreamGenerator {
+    private static final String DEBUG_NAME = "FixedWidthChunkInputStreamGenerator";
+
+    @FunctionalInterface
+    public interface TypeConversion<T> {
+        T apply(DataInput in) throws IOException;
+    }
+
+    /**
+     * Generic input stream reading from arrow's buffer and convert directly to java type.
+     *
+     * If useDeephavenNulls is enabled, then the conversion method must properly return a null value.
+     *
+     * @param elementSize    the number of bytes per element (element size is fixed)
+     * @param options        the stream reader options
+     * @param conversion     the conversion method from input stream to the result type
+     * @param fieldNodeIter  arrow field node iterator
+     * @param bufferInfoIter arrow buffer info iterator
+     * @param is             data input stream
+     * @param <T>            the result type
+     * @return               the resulting chunk of the buffer that is read
+     */
+    static <T> ObjectChunk<T, Values> extractChunkFromInputStreamWithTypeConversion(
+            final int elementSize,
+            final StreamReaderOptions options,
+            final TypeConversion<T> conversion,
+            final Iterator<ChunkInputStreamGenerator.FieldNodeInfo> fieldNodeIter,
+            final TLongIterator bufferInfoIter,
+            final DataInput is) throws IOException {
+
+        final ChunkInputStreamGenerator.FieldNodeInfo nodeInfo = fieldNodeIter.next();
+        final long validityBuffer = bufferInfoIter.next();
+        final long payloadBuffer = bufferInfoIter.next();
+
+        final WritableObjectChunk<T, Values> chunk = WritableObjectChunk.makeWritableChunk(nodeInfo.numElements);
+
+        if (nodeInfo.numElements == 0) {
+            return chunk;
+        }
+
+        final int numValidityLongs = options.useDeephavenNulls() ? 0 : (nodeInfo.numElements + 63) / 64;
+        try (final WritableLongChunk<Values> isValid = WritableLongChunk.makeWritableChunk(numValidityLongs)) {
+            if (options.useDeephavenNulls() && validityBuffer != 0) {
+                throw new IllegalStateException("validity buffer is non-empty, but is unnecessary");
+            }
+            int jj = 0;
+            for (; jj < Math.min(numValidityLongs, validityBuffer / 8); ++jj) {
```

Comment:
I have a (maybe superstitious) belief that it's better to write down the bounds when there's any computation to be done.
```suggestion
            final int numValidityLongsPresent = Math.min(numValidityLongs, validityBuffer / 8);
            for (; jj < numValidityLongsPresent; ++jj) {
```